pkg/grpc/tls/provider.go

```diff
@@ -27,25 +30,79 @@ type CertificateProvider interface {
 
 	// GetClientCertificate returns the current certificate to present to the server.
 	GetClientCertificate(*tls.CertificateRequestInfo) (*tls.Certificate, error)
+}
 
-	// UpdateCertificate updates the stored certificate for the given client request.
-	UpdateCertificates([]byte, *tls.Certificate) error
+// Generator describes an interface to sign the CSR.
+type Generator interface {
+	Identity(csr *talosx509.CertificateSigningRequest) (ca, crt []byte, err error)
 }
 
-type embeddableCertificateProvider struct {
+type certificateProvider struct {
 	sync.RWMutex
 
+	generator Generator
+
 	ca  []byte
 	crt *tls.Certificate
 
 	dnsNames []string
 	ips      []net.IP
+}
+
+// NewRenewingCertificateProvider returns a new CertificateProvider
+// which manages and updates its certificates using Generator.
+func NewRenewingCertificateProvider(generator Generator, dnsNames []string, ips []net.IP) (CertificateProvider, error) {
+	provider := &certificateProvider{
+		generator: generator,
+		dnsNames:  dnsNames,
+		ips:       ips,
+	}
+
+	var (
+		ca   []byte
+		cert tls.Certificate
+		err  error
+	)
+
+	if ca, cert, err = provider.update(); err != nil {
+		return nil, fmt.Errorf("failed to create initial certificate: %w", err)
+	}
+
+	provider.updateCertificates(ca, &cert)
+
+	// nolint: errcheck
+	go provider.manageUpdates(context.Background())
```

Comment:
This is exactly the kind of place `context.TODO()` was intended for.

Reply:
Because ultimately, NewRenewingCertificateProvider should accept a context.  Probably out of scope for this PR, but important to have static code analysis cue off this error.  Goroutine spawned with no termination mechanism.

Reply:
agree, I'll get it updated, but this code was copy-pasted as I squashed two providers together.

I'm not sure if that's the best way as context doesn't allow to wait for provider to stop, we can only cancel it. But that might be something for the future PR.

Reply:
I'm not sure I follow the second part.  Isn't this what you are talking about?

```go
for {
  select {
    case <-ctx.Done():
      return
    default:
  }
}
```

Reply:
I mean the caller (whoever creates the context) can cancel the context, but it can't wait for the provider to actually stop. So sometimes the goal is to make sure the provider fully stopped doing whatever it was busy with, and that is not possible with the context alone, we need some kind of `Wait()` mechanism. 

Right now we have providers used only in `main.go` of two services (`apid` and `trustd` iirc), so this isn't much of an issue.

Reply:
Ah, yeah, okay.  True, that mechanism would be _more_ useful.  I'm just mainly concerned about an un-cleaned-up goroutine... even if it should generally have the lifecycle of the process itself.

But not enough to block.  I think that changing it to `context.TODO()` is more than sufficient at the moment.